lib/client/kubesession.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/client/terminal"
+	"github.com/gravitational/teleport/lib/kube/proxy/streamproto"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+const mfaChallengeInterval = time.Second * 30
+
+// KubeSession a joined kubernetes session from the client side.
+type KubeSession struct {
+	stream    *streamproto.SessionStream
+	term      *terminal.Terminal
+	ctx       context.Context
+	closeCtx  context.CancelFunc
+	closeWait *sync.WaitGroup
+	meta      types.SessionTracker
+}
+
+// NewKubeSession joins a live kubernetes session.
+func NewKubeSession(ctxBg context.Context, tc *TeleportClient, meta types.SessionTracker, key *Key, kubeAddr string, tlsServer string, mode types.SessionParticipantMode) (*KubeSession, error) {
```

Comment:
```suggestion
func NewKubeSession(ctx context.Context, tc *TeleportClient, meta types.SessionTracker, key *Key, kubeAddr string, tlsServer string, mode types.SessionParticipantMode) (*KubeSession, error) {
```

Stick with `ctx` as the variable name here.